File: crm/src/types/customer.ts
export type PaymentStatus = "pending" | "partial" | "paid";
export type Gender = "male" | "female" | "other";

export type Customer = {
  id: number;
  customer_id: string;
  name: string;
  age: number | null;
  contact_no: string;
  email: string | null;
  whatsapp_no: string | null;
  gender: Gender | null;
  occupation: string | null;
  guardian_name: string | null;
  guardian_contact_no: string | null;
  address: string | null;
  purpose_of_visit: string | null;
  whatsapp_opt_in: boolean;
  created_at: string;
  updated_at: string;
  created_by: number | null;
  updated_by: number | null;
  is_deleted: boolean;
};

export type CustomerPrescriptionSummary = {
  id: number;
  prescription_date: string;
  notes: string | null;
};

export type CustomerBillSummary = {
  id: number;
  bill_number: string;
  final_price: number;
  balance_amount: number;
  payment_status: PaymentStatus;
  created_at: string;
};

export type CustomerVisitSummary = {
  id: number;
  visit_date: string;
  reason_for_visit: string;
  referred_by: string | null;
  status: "draft" | "in_progress" | "completed" | "cancelled";
  assigned_examiner_id: number | null;
  visit_notes: string | null;
  created_at: string;
};

export type CustomerReferralSummary = {
  visit_id: number;
  visit_date: string;
  specialist_type: string | null;
  referral_status: string | null;
  notes: string | null;
  follow_up: string | null;
};

export type CustomerContactLensOrderSummary = {
  id: number;
  visit_id: number;
  order_reference: string;
  status: "draft" | "ready_for_vendor" | "sent_to_vendor" | "in_production" | "ready_for_delivery" | "delivered" | "cancelled";
  vendor_id: number | null;
  delivered_by: number | null;
  delivered_at: string | null;
  created_at: string;
};

export type CustomerDispensingOrderSummary = CustomerContactLensOrderSummary;

export type CustomerFollowUpTaskSummary = {
  id: number;
  visit_id: number;
  contact_lens_order_id: number | null;
  task_type: string;
  due_date: string;
  status: "pending" | "completed" | "cancelled";
  notes: string | null;
  assigned_staff_id: number | null;
  reminder_state: "not_scheduled" | "scheduled" | "sent" | "failed";
  completion_notes: string | null;
  completed_at: string | null;
};

export type CustomerTimelineItem = {
  event: string;
  occurred_at: string;
  label: string;
  visit_id: number | null;
  entity_type: string;
  entity_id: number;
  status: string | null;
  user_id: number | null;
  notes: string | null;
  previous_status: string | null;
};

export type CustomerDetail = Customer & {
  visits: CustomerVisitSummary[];
  referrals: CustomerReferralSummary[];
  prescriptions: CustomerPrescriptionSummary[];
  bills: CustomerBillSummary[];
  dispensing_orders: CustomerDispensingOrderSummary[];
  contact_lens_orders: CustomerContactLensOrderSummary[];
  follow_up_tasks: CustomerFollowUpTaskSummary[];
  timeline: CustomerTimelineItem[];
};

export type CustomerPayload = {
  name: string;
  age?: number | null;
  contact_no: string;
  email?: string | null;
  whatsapp_no?: string | null;
  gender?: Gender | null;
  occupation?: string | null;
  guardian_name?: string | null;
  guardian_contact_no?: string | null;
  address?: string | null;
  purpose_of_visit?: string | null;
  whatsapp_opt_in: boolean;
  registration_idempotency_key?: string | null;
};

File: backend/app/schemas/customer.py
from __future__ import annotations

import re
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.enums import DispensingOrderStatus, FollowUpStatus, Gender, PaymentStatus

EMAIL_REGEX = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")


def _blank_to_none(value: object) -> object:
    if isinstance(value, str) and not value.strip():
        return None
    return value


class CustomerBase(BaseModel):
    name: str = Field(min_length=2, max_length=255)
    age: int | None = Field(default=None, ge=0, le=130)
    contact_no: str = Field(min_length=8, max_length=20)
    email: str | None = Field(default=None, max_length=255)
    whatsapp_no: str | None = Field(default=None, min_length=8, max_length=20)
    gender: Gender | None = None
    occupation: str | None = Field(default=None, max_length=255)
    guardian_name: str | None = Field(default=None, max_length=255)
    guardian_contact_no: str | None = Field(default=None, min_length=8, max_length=20)
    address: str | None = Field(default=None, max_length=2000)
    purpose_of_visit: str | None = Field(default=None, max_length=255)
    whatsapp_opt_in: bool = False

    @field_validator("whatsapp_no", "occupation", "guardian_name", "guardian_contact_no", "address", "purpose_of_visit", mode="before")
    @classmethod
    def normalize_blank_optional_strings(cls, value: object) -> object:
        return _blank_to_none(value)

    @field_validator("email")
    @classmethod
    def validate_email(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip().lower()
        if not normalized:
            return None
        if not EMAIL_REGEX.match(normalized):
            raise ValueError("Enter a valid email address")
        return normalized


class CustomerCreate(CustomerBase):
    registration_idempotency_key: str | None = Field(default=None, min_length=8, max_length=120)


class CustomerUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=255)
    age: int | None = Field(default=None, ge=0, le=130)
    contact_no: str | None = Field(default=None, min_length=8, max_length=20)
    email: str | None = Field(default=None, max_length=255)
    whatsapp_no: str | None = Field(default=None, min_length=8, max_length=20)
    gender: Gender | None = None
    occupation: str | None = Field(default=None, max_length=255)
    guardian_name: str | None = Field(default=None, max_length=255)
    guardian_contact_no: str | None = Field(default=None, min_length=8, max_length=20)
    address: str | None = Field(default=None, max_length=2000)
    purpose_of_visit: str | None = Field(default=None, max_length=255)
    whatsapp_opt_in: bool | None = None

    @field_validator("whatsapp_no", "occupation", "guardian_name", "guardian_contact_no", "address", "purpose_of_visit", mode="before")
    @classmethod
    def normalize_blank_optional_strings(cls, value: object) -> object:
        return _blank_to_none(value)

    @field_validator("email")
    @classmethod
    def validate_email(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip().lower()
        if not normalized:
            return None
        if not EMAIL_REGEX.match(normalized):
            raise ValueError("Enter a valid email address")
        return normalized


class CustomerRead(CustomerBase):
    model_config = ConfigDict(from_attributes=True)

    id: int
    customer_id: str
    created_at: datetime
    updated_at: datetime
    created_by: int | None
    updated_by: int | None
    is_deleted: bool


class CustomerPrescriptionSummary(BaseModel):
    id: int
    prescription_date: date
    notes: str | None


class CustomerBillSummary(BaseModel):
    id: int
    bill_number: str
    final_price: Decimal
    balance_amount: Decimal
    payment_status: PaymentStatus
    created_at: datetime


class CustomerVisitSummary(BaseModel):
    id: int
    visit_date: datetime
    reason_for_visit: str
    referred_by: str | None
    status: str
    assigned_examiner_id: int | None
    visit_notes: str | None
    created_at: datetime


class CustomerReferralSummary(BaseModel):
    visit_id: int
    visit_date: datetime
    specialist_type: str | None
    referral_status: str | None
    notes: str | None
    follow_up: str | None


class CustomerContactLensOrderSummary(BaseModel):
    id: int
    visit_id: int
    order_reference: str
    status: DispensingOrderStatus
    vendor_id: int | None
    created_at: datetime


class CustomerFollowUpTaskSummary(BaseModel):
    id: int
    visit_id: int
    contact_lens_order_id: int
    task_type: str
    due_date: date
    status: FollowUpStatus
    notes: str | None
    completed_at: datetime | None


class CustomerDetailRead(CustomerRead):
    visits: list[CustomerVisitSummary]
    referrals: list[CustomerReferralSummary]
    prescriptions: list[CustomerPrescriptionSummary]
    bills: list[CustomerBillSummary]
    contact_lens_orders: list[CustomerContactLensOrderSummary]
    follow_up_tasks: list[CustomerFollowUpTaskSummary]


class CustomerListResponse(BaseModel):
    items: list[CustomerRead]
    total: int
    page: int
    page_size: int
